### DISCUSSION TITLE
Reaction forces with `sympy.physics.mechanics.System` and `KanesMethod`

### DISCUSSION THREAD

**Discussion by Davide-sd** [sympy/sympy#27011] [General]
@tjstienstra @moorepants 

I'm trying the `System` class introduced with sympy 1.13.0. However, I'm unable to figure it out how to compute reaction forces.

I'm going to use the simple pendulum as an example. In the following diagram, `O` is the origin, `P` is the particle of mass `m`, while `L` is the distance between the particle and the origin. I'd like to compute the equation of motions of the system, as well as the reaction forces acting on the origin.

![image](https://github.com/user-attachments/assets/d69eee8f-8dd0-4735-80c9-a43f08c4b985)

```py
from sympy import *
import sympy.physics.mechanics as me
me.init_vprinting()

theta, omega = me.dynamicsymbols("theta, omega")
m, g, L = symbols("m, g, L")

# reaction forces acting on point O
Rx, Ry, Rz = me.dynamicsymbols("R_x, R_y, R_z")
# auxiliary velocities needed to compute the reaction forces
u1_aux, u2_aux, u3_aux = me.dynamicsymbols("u_aux^((1:4))")

N, A = symbols("N, A", cls=me.ReferenceFrame)

wall = me.RigidBody("Wall", frame=N)
particle = me.Particle("Particle")

# set fictitious  velocity to point O
wall.masscenter.set_vel(N, u1_aux * N.x + u2_aux * N.y + u3_aux * N.z)
print("Before adding joints:")
print("\tVelocity of O in N:", wall.masscenter.vel(N))

system = me.System.from_newtonian(wall)
system.add_auxiliary_speeds(u1_aux, u2_aux, u3_aux)
system.add_joints(
    me.PinJoint(
        "J1", wall, particle, coordinates=theta, speeds=omega,
        child_interframe=A,
        child_point=-L * A.y,
        joint_axis=wall.frame.z
    )
)
print("After adding joints:")
print("\tVelocity of O in N:", wall.masscenter.vel(N))
print("\tVelocity of P in N:", particle.masscenter.vel(N))

system.apply_uniform_gravity(g * system.y)
system.add_loads(
    # reaction forces acting on O
    me.Force(wall.masscenter, Rx * N.x + Ry * N.y + Rz * N.z)
)
system.validate_system()
eoms = system.form_eoms()
eoms
```
![image](https://github.com/user-attachments/assets/c0eb04f1-9058-48e1-90db-21ef4ae52228)

The equation of motion is correct, but the auxiliary equations are all set to zero, which is wrong. Note that I set the fictitious velocity of the origin before creating the pin joint. However, the pin joint constructor is going to modify that velocity and set it to zero. Which means, the auxiliary speeds are not propagated to the velocity of the particle.

Am I doing something wrong or is this a bug?

**Comment by tjstienstra** [comment]
@Davide-sd Good find, I would indeed flag this as a bug. The problem indeed seems to be in the joints framework, which doesn't take into account that you may have applied auxiliary velocities. It just sets the velocities of the intermediate points to zero (e.g. [here](https://github.com/sympy/sympy/blob/01a25577c8acda8f50a7d35eb0971fc0a1e17344/sympy/physics/mechanics/joint.py#L816)).

I would have to check what the most optimal fix would be in the source code (the second would probably have my preference):
- Remove specifying does velocities to be zero and use the velocity theories. Would have to investigate what is the influence of this change.
- Check if the velocity of the intermediate point has not yet been specified. If it hasn't then specify it based on the masscenter of the respective body.

In your case the easiest fix would be to apply the auxiliary speeds after defining the initial kinematics, e.g.
```python
# Define kinematics/joints etc.
...
# Apply auxiliary speeds
aux_vel = u1_aux * N.x + u2_aux * N.y + u3_aux * N.z
wall.masscenter.set_vel(N, aux_vel)
particle.masscenter.set_vel(N, particle.masscenter.vel(N) + aux_vel)
```
This idea of applying the speeds afterwards is one I've used in [BRiM](https://github.com/mechmotum/brim), a package to derive the EoMs of bicycle-rider models using SymPy. [Here](https://github.com/mechmotum/brim/blob/main/src/brim/core/auxiliary.py) I've written a class that does the propagation for you based on how you defined points w.r.t. each other in terms of positions.
